### PR TITLE
vsock_proxy: Perform DNS resolution after the expiration of the TTL

### PR DESCRIPTION
**Pull request by meerd**
**Description of changes:** This PR addresses the wrong assumption that domain names are resolved to fixed static IP addresses. With these changes, **vsock-proxy** periodically refreshes the resolved IP address of the remote host after the Time-To-Live (TTL) expires, ensuring that the most up-to-date IP address is used.

**Relevant issues:** #553 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: vsock_proxy/Cargo.toml [hunk 1]
```diff
@@ -12,6 +12,7 @@ chrono = "0.4"
 clap = "3.2"
 dns-lookup = "2.0.3"
```
Comment (foersleo) [inline vsock_proxy/Cargo.toml:13]: Looks like `dns-lookup` is not used anywhere anymore. We should probably remove it here as well.

File: vsock_proxy/src/dns.rs [hunk 2]
```diff
@@ -18,35 +18,42 @@ pub fn resolve(addr: &str, ip_addr_type: IpAddrType) -> VsockProxyResult<Vec<Dns
 
     // DNS lookup
     // It results in a vector of IPs (V4 and V6)
-    let ips = lookup_host(&addr).map_err(|_| "DNS lookup failed!")?;
+    let resolver = Resolver::new(ResolverConfig::default(), ResolverOpts::default()).unwrap();
+
+    let rresults: Vec<DnsResolveResult> = resolver
+        .lookup_ip(addr)
+        .map_err(|error| format!("DNS lookup failed! {:?}", error))?
```
Comment (foersleo) [inline vsock_proxy/src/dns.rs:25]: This changes the format of the error message when the lookup fails. This makes `test_resolve_invalid_domain` unhappy as it matches on `DNS lookup failed!`. So the test probably needs an update to account for the extra info about the error type. I believe this is what makes the integration tests fail.

**Review by foersleo — COMMENTED** [review]
Apart from the comments about the mismatch in error message and test pointed out before this looks good to me.

File: vsock_proxy/src/dns.rs (hunk 2 above)
Comment (meerd) [reply vsock_proxy/src/dns.rs]: Nice catch! Fixed in the new revision (1415a7d).

**Review by foersleo — APPROVED** [review]

File: vsock_proxy/src/dns.rs [hunk 3]
```diff
@@ -12,48 +12,50 @@ use std::net::IpAddr;
...
-        .map_err(|err| format!("Could not parse remote address: {}", err))?;
-
-    let ip = *addrs.first().ok_or("No IP address found")?;
-    let ttl = 60; // IMPORTANT TODO: Obtain this value dynamically
-
-    Ok(DnsResolveResult { ip, ttl })
+    let rresults = resolve(addr, ip_addr_type)?;
+    // Return the first resolved IP address and its TTL value.
+    // We already know that rresults containt at least one item.
+    Ok(*rresults.first().unwrap())
```
Comment (mariusknaust) [inline vsock_proxy/src/dns.rs:67]: We should be able to avoid the unwrap by `rresult.first().ok_or(...)`

File: vsock_proxy/src/dns.rs [hunk 4]
```diff
@@ -18,35 +18,42 @@ pub fn resolve(addr: &str, ip_addr_type: IpAddrType) -> VsockProxyResult<Vec<Dns
 
     // DNS lookup
     // It results in a vector of IPs (V4 and V6)
-    let ips = lookup_host(&addr).map_err(|_| "DNS lookup failed!")?;
+    let resolver = Resolver::new(ResolverConfig::default(), ResolverOpts::default()).unwrap();
```
Comment (mariusknaust) [inline vsock_proxy/src/dns.rs:21]: We should use `map_err()` instead of unwrap here.

File: vsock_proxy/src/proxy.rs [hunk 5]
```diff
@@ -64,11 +64,23 @@ pub fn check_allowlist(
...
-            let rresults = dns::resolve(addr, ip_addr_type)?;
-            for rresult in rresults.into_iter() {
-                if rresult.ip == remote_addr {
-                    info!("Matched with host IP \"{}\" and port \"{}\"", addr, port);
-                    return Ok(remote_addr);
+            let rresults = dns::resolve(addr, ip_addr_type);
+            let remote_addr_matched = rresults.clone().ok().and_then(|rresults| {
+                rresults
+                    .into_iter()
+                    .find(|rresult| rresult.ip == remote_addr)
```
Comment (mariusknaust) [inline vsock_proxy/src/proxy.rs:71]: We should be able to simplify this by: `rresult.iter().flatten().find(...)`

**Review by mariusknaust — APPROVED** [review]

**Pull request merged**